### PR TITLE
[IMP] web_tour, *: remove default click trigger on final tour step

### PR DESCRIPTION
**Pull request by divy-odoo**
When we haven't provided a custom action, the tour step runs the default
action. In the final step of the tour, when there is no `run` or
`isCheck` provided, we don't want to execute the default action as it
can lead to a race condition.

With this commit [1], from saas-16.3 we have refactored the web_tour to
OWL. During the refactoring, it lost the condition initially introduced
by PR[2] that handles the final step of tours.

So this PR,
1. Restored the missing condition.
2. Replaced the simple `console.log()` with a warning, because it's easy
   to miss the log compared to a warning.
3. Rsolve the warnings raised by the last tour steps

[1]: https://github.com/odoo/odoo/commit/3a798039d6f200f8e28448ddb6a2d3c46654a203
[2]: https://github.com/odoo/odoo/commit/9733f6e1ab31abf780a5ad671f700fac883b63ee

task-3429500

PR Enterprise: https://github.com/odoo/enterprise/pull/46683

### CODE REVIEW

File: addons/website/static/tests/tours/website_form_editor_frontend.js
Comment (divy-odoo) [inline addons/website/static/tests/tours/website_form_editor_frontend.js:151]: IDK if this double space is intended, should we remove that?

File: addons/point_of_sale/static/tests/tours/helpers/ProductScreenTourMethods.js
Comment (rdeodoo) [inline addons/point_of_sale/static/tests/tours/helpers/ProductScreenTourMethods.js:410]: Remove this, it's the default behavior

File: addons/project/static/tests/tours/project_sharing_tour.js
Comment (rdeodoo) [inline addons/project/static/tests/tours/project_sharing_tour.js:91]: remove this
You can either:
- Remove this line and add a new step afterward which correctly checks that the "Go to list view" correctly behaved: just a check step about list view being loaded
- Or simply remove this step, it's probably doing nothing: it clicks and directly stops the test without waiting anything (which lead to race condition generally, because it stops the second it clicks on the button, then the tour manager is killing everything but there is still pending request which can't be handled correctly).

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thanks for the adaptation :) 

Just remove all the `run: 'click',` as last step, that the point of the task.
Such cases never make sense, it's an error during development.
If you click on something as last step, you need to check something afterward: did the page load correctly ? Was the action handled correctly ? Did the element got removed correctly  etc etc

File: addons/website/static/tests/tours/website_form_editor_frontend.js
Comment (rdeodoo) [reply addons/website/static/tests/tours/website_form_editor_frontend.js:151]: It was a mistake and full of linter error but keep it like that, it would need a separate commit to fix those linter error so it's not noise in your pr

**Comment by divy-odoo** [comment]
Hello @rdeodoo, Changes have been made.

-> Removed `run: 'click'` and added a few check steps instead.

File: addons/website_forum/static/src/js/tours/website_forum.js
Comment (rdeodoo) [inline addons/website_forum/static/src/js/tours/website_forum.js:62]: Don't change the wording of the step, this is visible to users through real tour (not only test).
I am not sure what would be the best solution for those tours requiring a user click as last step.
I guess it should be about adding a next step which would be run only in test mode and not "user mode" with `auto: true`.
In this case, restore the step that you changed, and add a new step after which would check that the "Click here to accept this answer" ran correctly.

```js
{
    trigger: ".something-appearing-after-accepting-the-answer"
    auto: true,
    isCheck: true,
}
```

File: addons/website_blog/static/src/js/tours/website_blog.js
Comment (rdeodoo) [inline addons/website_blog/static/src/js/tours/website_blog.js:91]: See review point `4.`, don't change this step but add a new one (check:true) for test only (auto:true)

File: addons/mail/static/src/discuss/core/web/discuss_tour.js
Comment (rdeodoo) [inline addons/mail/static/src/discuss/core/web/discuss_tour.js:90]: See review point `4.`, don't change this step but add a new one (check:true) for test only (auto:true)

File: addons/project/static/tests/tours/project_burndown_chart_tour.js
Comment (rdeodoo) [inline addons/project/static/tests/tours/project_burndown_chart_tour.js:76]: This step is actually useless :thinking: `:not(:has(.o_comparison_menu))` will always be true, or is there something I am missing? Anyway, it's not related to your PR, let's ignore that

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thank you for your work, it must not have been a funny one.

I am setting the PR as `ready for review` so people can jump in should they want to.

1. You mentioned in your commit message the [commit](https://github.com/odoo/odoo/commit/3a798039d6f200f8e28448ddb6a2d3c46654a203) which broke the behavior but not the [commit](https://github.com/odoo/odoo/commit/9733f6e1ab31abf780a5ad671f700fac883b63ee) which initially introduced the behavior which is restored (and made blocking rather than informative) here.
Could you mention it?
2. It seems like you forgot the `design-themes` and `enterprise` repository? I didn't see any build where those were built (I've just build those on the last build).
3. Could you mention in your commit message that on top of possibly being race condition prone (which you wrote :+1: ), it's also generally a mistake that the developer didn't see and wrote something which is not doing what they actually wanted to do? Which this PR proves given the amount of click being done on non clickable elements.
4. For tours (not tests) requiring a user click as last step, don't change that step but add a new step with `auto:true, isCheck: true` which will check that the (previous) last step click ran correctly. I made an example in one of the review point.

Remarks:
- the first implementation (restored here) was a simple console.log which actually made people waste some time as it was unclear why their test was not working as expected during development. The only way to understand what was going on was to not miss the log entry.
- Some steps in tours are missing `isCheck: true` here and there, but this PR is just about the last step ones.
- see https://discord.com/channels/678381219515465750/687338039717920792/1139215709856079965 and https://discord.com/channels/678381219515465750/687338039717920792/1141300909712887828 for exchange on framework-js discord about this PR

File: addons/website_forum/static/src/js/tours/website_forum.js
Comment (divy-odoo) [reply addons/website_forum/static/src/js/tours/website_forum.js:62]: Hi @rdeodoo ,

Actually, Performing the last step: `_t("Click here to accept this answer.")` produces a warning on the UI side (see screenshot). I thought it was actually expected behaviour So I changed the last step instead of adding a check for a warning. But I was wrong there, I just researched more and it's working fine for 16.3 and below. So I guess it's an issue, a warning should not be there.

**Issue**: IMO, Users should able to make their own answers as best on their own Questions (same as 16.3 and below).

Like to have your view on it, is it expected behaviour or not? If not, should we wait for the fix?

Questions

File: addons/hr_holidays/static/src/tours/hr_holidays_tour.js
Comment (divy-odoo) [reply addons/hr_holidays/static/src/tours/hr_holidays_tour.js:101]: Hello @rdeodoo,

The tour is about approving a Time off request, After performing the last step there is nothing special left on the screen to check. I found this the best way to handle this so I just added these steps to just check if my leave is approved.

Before last step : `content: _t("Let's approve it"),`
Timeoff_approval

After the last step performed : 
After_approval

Comment (rdeodoo) [reply addons/hr_holidays/static/src/tours/hr_holidays_tour.js:101]: It's not the scope of this task to alter existing tours, let's not do that.

> there is nothing special left on the screen to check.

You could very well check that the breadcrumb status moved to "Approved" as last step, and make it `auto: true` so it's only a step used in tests and not shown to the user.
image

File: addons/website_forum/static/src/js/tours/website_forum.js
Comment (rdeodoo) [reply addons/website_forum/static/src/js/tours/website_forum.js:62]: I don't know, it's probably a mistake, but is it relevant to this PR?
At best, you could submit a feedback about it to the related team.

Comment (divy-odoo) [reply addons/website_forum/static/src/js/tours/website_forum.js:62]: Hello @rdeodoo, A [task](https://www.odoo.com/web#id=3481264&cids=2&model=project.task&view_type=form) has been created for the issue. An error message is intended but the message should be something else as mentioned in the task.

> but is it relevant to this PR?

Yes, It's about the last tour step which ignored the action to perform.

Should we wait for the fix or go ahead by just checking the warning( ignoring the error message that appears )?

Comment (flch-odoo) [reply addons/website_forum/static/src/js/tours/website_forum.js:62]: Hello,
The error is not a mistake, we did not want to allow this from 16.4 onward. The tmentionned ask will rephrase the error message.

This step could indeed be updated to simply check that the answer is present, but what would make be even better is checking that the button to edit the answer is present, if we may ask this of you while you are here?

In any case, thank you for the care put into this!

File: addons/hr_holidays/static/src/tours/hr_holidays_tour.js
Comment (rdeodoo) [reply addons/hr_holidays/static/src/tours/hr_holidays_tour.js:101]: > For that too, I have to write 2-3 extra tour steps, will that be okay?

No, we don't want to add any step other than a real `isCheck:true` and `auto:true` in this PR.
The PR should just be about either add the missing `isCheck:true` or adding a check step when the last step was really an action.

In this case, I don't understand why don't you just add a check step after the last action which is clicking on approve.
Your diff in this file should just be about adding a step the check if the entry in the list view is gone.
Indeed, after you approve the time off, once the RPC is sent and get back, the entry is removed from the list view, you can then just write a selector in the form of `body:not(:has(.some_list_view_entry_selector:contains('Something Unique To The Entry That You Approve')))`.
The unique thing could be what you think is the more robust: one of the field, or a combination of multiple fields, or the button "Approve". Basically anything that is on screen before you click and not on screen after you have clicked. The step will then be falsy until the element is hidden, meaning all RPC went back and got processed.

**Comment by rdeodoo** [comment]
You will need to rebase the branch, there is 2 conflicts.
Ideally don't wait too much as this will not be fun to maintain, so we can ship it quickly :+1:

**Comment by divy-odoo** [comment]
Hello @rdeodoo, PR is ready for the final review.

Enterprise PR: https://github.com/odoo/enterprise/pull/46683 ( Ready for the review )

File: addons/hr_holidays/static/src/tours/hr_holidays_tour.js
Comment (rdeodoo) [inline addons/hr_holidays/static/src/tours/hr_holidays_tour.js:89]: :detective: Typo
And the translation is actually useless on an `auto: true` step since it's only run in tests, not shown to users.
Arguably you could even omit the `content` property but I prefer to keep it, it's easier when debugging tours.

File: addons/website_forum/static/src/js/tours/website_forum.js
Comment (rdeodoo) [reply addons/website_forum/static/src/js/tours/website_forum.js]: image
It's not really what was asked here.. I'll push force the last changes directly so this can be merged

**Comment by rdeodoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**
